File: frontend/src/types/storms.ts
import type { GeoJsonObject } from 'geojson'

export interface StormForecastPoint {
  lat: number
  lon: number
  eta_minutes: number
  label: string
}

export interface StormImpact {
  location_id: string
  location_name: string
  eta_minutes_low: number | null
  eta_minutes_high: number | null
  distance_km: number | null
  threat_at_arrival: string
  trend_at_arrival: string
  confidence: number
  summary: string
  impact_rank: number
  details: Record<string, unknown>
}

export interface StormEnvironment {
  source?: string | null
  current_station_id?: string | null
  future_station_id?: string | null
  gridpoint_id?: string | null
  surface_temp_c?: number | null
  dewpoint_c?: number | null
  wind_speed_kt?: number | null
  forecast_probability_of_thunder?: number | null
  ahead_probability_of_thunder?: number | null
  forecast_qpf_mm?: number | null
  forecast_wind_speed_kmh?: number | null
  hail_favorability?: number | null
  wind_favorability?: number | null
  tornado_favorability?: number | null
  heavy_rain_favorability?: number | null
  convective_signal?: number | null
  intensification_signal?: number | null
  weakening_signal?: number | null
  projected_trend?: string | null
  projection_confidence?: number | null
  environment_confidence?: number | null
  environment_freshness_minutes?: number | null
  environment_ahead_delta?: Record<string, number | null>
  cape_jkg?: number | null
  cin_jkg?: number | null
  bulk_shear_06km_kt?: number | null
  bulk_shear_01km_kt?: number | null
  srh_surface_925hpa_m2s2?: number | null
  dcape_jkg?: number | null
  dcape_is_proxy?: boolean | null
  freezing_level_m?: number | null
  pwat_mm?: number | null
  lapse_rate_midlevel_cpkm?: number | null
  lcl_m?: number | null
  lfc_m?: number | null
  model_valid_at?: string | null
  ahead_model_valid_at?: string | null
  weather_summary?: string | null
  hazards?: string[]
  ahead_trend?: string | null
  limitation?: string | null
  profile_summary?: Record<string, unknown>
  field_provenance?: Record<string, string>
  source_notes?: string[]
  hodograph?: Record<string, unknown>
  srv_metrics?: Record<string, unknown>
  operational_context?: Record<string, unknown>
  volume_metrics?: Record<string, unknown>
}

export interface StormPrediction {
  intensification_score?: number | null
  weakening_score?: number | null
  maintenance_score?: number | null
  projected_trend?: string | null
  projected_primary_threat?: string | null
  projected_secondary_threats?: string[]
  projected_confidence?: number | null
  projected_threat_scores?: Record<string, number>
  forecast_reasoning_factors?: string[]
  environment_confidence?: number | null
  motion_confidence?: number | null
  persistence_score?: number | null
  forecast_stability_score?: number | null
  data_quality_score?: number | null
  uncertainty_factors?: string[]
}

export interface UncertaintyConeStep {
  eta_minutes: number
  center: { lat: number; lon: number }
  left: { lat: number; lon: number }
  right: { lat: number; lon: number }
  half_width_km: number
}

export interface StormSummary {
  storm_id: string
  site: string
  latest_frame_id: string | null
  latest_scan_time: string
  created_at?: string | null
  updated_at?: string | null
  status: string
  lifecycle_state: string
  centroid_lat: number
  centroid_lon: number
  area_km2: number
  max_reflectivity: number
  mean_reflectivity: number
  motion_heading_deg: number | null
  motion_speed_kmh: number | null
  trend: string
  primary_threat: string
  secondary_threats: string[]
  severity_level: string
  confidence: number
  threat_scores: Record<string, number>
  narrative: string
  reasoning_factors: string[]
  footprint_geojson: GeoJsonObject
  forecast_path: StormForecastPoint[]
  // v12 — uncertainty cone and convective mode
  uncertainty_cone: UncertaintyConeStep[]
  storm_mode: string
  storm_mode_confidence: number
  storm_mode_evidence: string[]
  track_uncertainty_km: number
  associated_signatures: Array<Record<string, unknown>>
  environment_summary: StormEnvironment | null
  prediction_summary: StormPrediction | null
  near_term_expectation: string
  impacts: StormImpact[]
  // v13 — threat component breakdown and lifecycle analysis
  threat_component_breakdown?: Record<string, Record<string, number>>
  threat_top_reasons?: Record<string, string[]>
  threat_limiting_factors?: Record<string, string[]>
  lifecycle_summary?: Record<string, unknown>
  // v14 — event flags and operational priority
  event_flags?: EventFlag[]
  priority_score?: number
  priority_label?: string
}

export interface StormTrackPoint {
  scan_time: string
  centroid_lat: number
  centroid_lon: number
  max_reflectivity: number
  mean_reflectivity: number
  motion_heading_deg: number | null
  motion_speed_kmh: number | null
  trend: string
}

export interface SavedLocation {
  location_id: string
  name: string
  lat: number
  lon: number
  kind: string
  created_at: string
  updated_at: string
}

export interface SavedLocationCreate {
  name: string
  lat: number
  lon: number
  kind?: string
}

export interface MetarObservation {
  station_id: string
  observation_time: string | null
  lat: number
  lon: number
  temp_c: number | null
  dewpoint_c: number | null
  wind_dir_deg: number | null
  wind_speed_kt: number | null
  wind_gust_kt: number | null
  visibility_mi: number | null
  pressure_hpa: number | null
  flight_category: string | null
  raw_text: string | null
  distance_km: number | null
}

export interface OverlayFeatureCollection {
  overlay_kind: string | null
  source: string | null
  type: 'FeatureCollection'
  fetched_at: string | null
  features: Array<Record<string, unknown>>
}

export interface EventFlag {
  flag: string
  label: string
  confidence: number
  rationale: string
  severity: number
  provenance: string
}

export interface StormHotspot {
  storm_id: string
  site: string
  priority_score: number
  priority_label: string
  severity_level: string
  primary_threat: string
  threat_scores: Record<string, number>
  storm_mode: string
  centroid_lat: number
  centroid_lon: number
  motion_heading_deg: number | null
  motion_speed_kmh: number | null
  confidence: number
  trend: string
  event_flags: EventFlag[]
  top_flag: string | null
  impact_count: number
  latest_scan_time: string | null
}

export interface LocationRiskEntry {
  location_id: string
  location_name: string
  lat: number
  lon: number
  risk_level: 'HIGH' | 'MODERATE' | 'LOW' | 'NONE'
  risk_score: number
  threatening_storm_count: number
  nearest_eta_low: number | null
  nearest_eta_high: number | null
  primary_threat: string | null
  threat_scores: Record<string, number>
  top_storm_id: string | null
  top_storm_severity: string | null
  top_impact_summary: string | null
  trend: string | null
  confidence: number | null
  event_flag_labels: string[]
}

export interface StormCompareField {
  label: string
  storm_a: unknown
  storm_b: unknown
  delta: number | null
  note: string | null
}

export interface StormCompareResponse {
  storm_a_id: string
  storm_b_id: string
  fields: StormCompareField[]
  provenance: string
}

export interface StormEventHistoryPoint {
  scan_time: string
  event_flags: EventFlag[]
  lifecycle_state: string | null
  priority_score: number | null
  priority_label: string | null
  severity_level: string | null
  primary_threat: string | null
  threat_scores: Record<string, number>
  storm_mode: string | null
  motion_heading_deg: number | null
  motion_speed_kmh: number | null
  confidence: number | null
}

export interface StormEventHistoryResponse {
  storm_id: string
  site: string
  point_count: number
  points: StormEventHistoryPoint[]
  provenance: string
}

export interface StormPrecomputedSummary {
  storm_id: string
  site: string
  computed_at: string
  scan_count: number
  first_seen: string | null
  last_seen: string | null
  peak_severity: string | null
  peak_threat_scores: Record<string, number>
  peak_reflectivity: number | null
  max_area_km2: number | null
  max_speed_kmh: number | null
  max_priority_score: number | null
  dominant_mode: string | null
  flag_summary: Array<{ flag: string; label: string; occurrence_count: number }>
  threat_trend: Array<{ scan_time: string; max_score: number; primary_threat: string }>
  motion_trend: Array<{ scan_time: string; speed_kmh: number | null; heading_deg: number | null }>
  impact_location_ids: string[]
  summary_narrative: string | null
  provenance: string
}

export interface SiteHistoryStatus {
  site: string
  last_ingest_time: string | null
  last_processing_cycle_time: string | null
  last_history_aggregation_time: string | null
  snapshot_count: number
  event_history_count: number
  precomputed_summary_count: number
  backlog_frame_count: number
  is_caught_up: boolean
  history_stale: boolean
}

File: backend/api/schemas/storms.py
from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field


class StormForecastPointResponse(BaseModel):
    lat: float
    lon: float
    eta_minutes: int
    label: str


class StormImpactResponse(BaseModel):
    location_id: str
    location_name: str
    eta_minutes_low: int | None
    eta_minutes_high: int | None
    distance_km: float | None
    threat_at_arrival: str
    trend_at_arrival: str
    confidence: float
    summary: str
    impact_rank: float
    details: dict[str, Any] = Field(default_factory=dict)


class StormEnvironmentResponse(BaseModel):
    source: str | None = None
    current_station_id: str | None = None
    future_station_id: str | None = None
    gridpoint_id: str | None = None
    surface_temp_c: float | None = None
    dewpoint_c: float | None = None
    wind_speed_kt: float | None = None
    forecast_probability_of_thunder: float | None = None
    ahead_probability_of_thunder: float | None = None
    forecast_qpf_mm: float | None = None
    forecast_wind_speed_kmh: float | None = None
    hail_favorability: float | None = None
    wind_favorability: float | None = None
    tornado_favorability: float | None = None
    heavy_rain_favorability: float | None = None
    convective_signal: float | None = None
    intensification_signal: float | None = None
    weakening_signal: float | None = None
    projected_trend: str | None = None
    projection_confidence: float | None = None
    environment_confidence: float | None = None
    environment_freshness_minutes: int | None = None
    environment_ahead_delta: dict[str, float | None] = Field(default_factory=dict)
    cape_jkg: float | None = None
    cin_jkg: float | None = None
    bulk_shear_06km_kt: float | None = None
    bulk_shear_01km_kt: float | None = None
    srh_surface_925hpa_m2s2: float | None = None
    dcape_jkg: float | None = None
    dcape_is_proxy: bool | None = None
    freezing_level_m: float | None = None
    pwat_mm: float | None = None
    lapse_rate_midlevel_cpkm: float | None = None
    lcl_m: float | None = None
    lfc_m: float | None = None
    model_valid_at: str | None = None
    ahead_model_valid_at: str | None = None
    weather_summary: str | None = None
    hazards: list[str] = Field(default_factory=list)
    ahead_trend: str | None = None
    limitation: str | None = None
    profile_summary: dict[str, Any] = Field(default_factory=dict)
    field_provenance: dict[str, str] = Field(default_factory=dict)
    source_notes: list[str] = Field(default_factory=list)
    hodograph: dict[str, Any] = Field(default_factory=dict)
    srv_metrics: dict[str, Any] = Field(default_factory=dict)
    operational_context: dict[str, Any] = Field(default_factory=dict)
    volume_metrics: dict[str, Any] = Field(default_factory=dict)


class StormPredictionResponse(BaseModel):
    intensification_score: float | None = None
    weakening_score: float | None = None
    maintenance_score: float | None = None
    projected_trend: str | None = None
    projected_primary_threat: str | None = None
    projected_secondary_threats: list[str] = Field(default_factory=list)
    projected_confidence: float | None = None
    projected_threat_scores: dict[str, float] = Field(default_factory=dict)
    forecast_reasoning_factors: list[str] = Field(default_factory=list)
    environment_confidence: float | None = None
    motion_confidence: float | None = None
    persistence_score: float | None = None
    forecast_stability_score: float | None = None
    data_quality_score: float | None = None
    uncertainty_factors: list[str] = Field(default_factory=list)


class StormSummaryResponse(BaseModel):
    storm_id: str
    site: str
    latest_frame_id: str | None
    latest_scan_time: datetime
    created_at: datetime | None = None
    updated_at: datetime | None = None
    status: str
    lifecycle_state: str
    centroid_lat: float
    centroid_lon: float
    area_km2: float
    max_reflectivity: float
    mean_reflectivity: float
    motion_heading_deg: float | None
    motion_speed_kmh: float | None
    trend: str
    primary_threat: str
    secondary_threats: list[str] = Field(default_factory=list)
    severity_level: str
    confidence: float
    threat_scores: dict[str, float] = Field(default_factory=dict)
    narrative: str
    reasoning_factors: list[str] = Field(default_factory=list)
    footprint_geojson: dict[str, Any]
    forecast_path: list[StormForecastPointResponse] = Field(default_factory=list)
    # v12 — uncertainty cone and convective mode
    uncertainty_cone: list[dict[str, Any]] = Field(default_factory=list)
    storm_mode: str = "unknown"
    storm_mode_confidence: float = 0.0
    storm_mode_evidence: list[str] = Field(default_factory=list)
    track_uncertainty_km: float = 5.0
    associated_signatures: list[dict[str, Any]] = Field(default_factory=list)
    environment_summary: StormEnvironmentResponse | None = None
    prediction_summary: StormPredictionResponse | None = None
    near_term_expectation: str = ""
    impacts: list[StormImpactResponse] = Field(default_factory=list)
    # v13 — threat component breakdown and lifecycle summary
    threat_component_breakdown: dict[str, dict[str, float]] = Field(default_factory=dict)
    threat_top_reasons: dict[str, list[str]] = Field(default_factory=dict)
    threat_limiting_factors: dict[str, list[str]] = Field(default_factory=dict)
    lifecycle_summary: dict[str, Any] = Field(default_factory=dict)
    # v14 — event flags and operational priority
    event_flags: list[dict[str, Any]] = Field(default_factory=list)
    priority_score: float = 0.0
    priority_label: str = "MINIMAL"


class StormTrackPointResponse(BaseModel):
    scan_time: datetime
    centroid_lat: float
    centroid_lon: float
    max_reflectivity: float
    mean_reflectivity: float
    motion_heading_deg: float | None
    motion_speed_kmh: float | None
    trend: str


class SavedLocationCreate(BaseModel):
    name: str = Field(min_length=1, max_length=80)
    lat: float = Field(ge=-90, le=90)
    lon: float = Field(ge=-180, le=180)
    kind: str = Field(default="custom", min_length=1, max_length=32)


class SavedLocationResponse(BaseModel):
    location_id: str
    name: str
    lat: float
    lon: float
    kind: str
    created_at: datetime
    updated_at: datetime


class MetarObservationResponse(BaseModel):
    station_id: str
    observation_time: str | None
    lat: float
    lon: float
    temp_c: float | None
    dewpoint_c: float | None
    wind_dir_deg: float | None
    wind_speed_kt: float | None
    wind_gust_kt: float | None
    visibility_mi: float | None
    pressure_hpa: float | None
    flight_category: str | None
    raw_text: str | None
    distance_km: float | None = None


class StormNarrativeResponse(BaseModel):
    storm_id: str
    narrative: str
    near_term_expectation: str
    confidence: float
    projected_confidence: float | None = None
    reasoning_factors: list[str] = Field(default_factory=list)
    forecast_reasoning_factors: list[str] = Field(default_factory=list)


class StormTimeSeriesPoint(BaseModel):
    """Single per-scan data point in a storm's time series."""
    scan_time: str
    centroid_lat: float
    centroid_lon: float
    area_km2: float
    max_reflectivity: float
    mean_reflectivity: float
    motion_speed_kmh: float | None = None
    motion_heading_deg: float | None = None
    trend: str
    severity_level: str
    confidence: float
    threat_scores: dict[str, float] = Field(default_factory=dict)


class StormTimeSeriesResponse(BaseModel):
    """Per-storm historical time series covering all tracked scans."""
    storm_id: str
    site: str
    point_count: int
    points: list[StormTimeSeriesPoint] = Field(default_factory=list)
    provenance: str = "Derived from stored storm-object snapshots. Each point represents one radar scan."


class ThreatComponentBreakdownResponse(BaseModel):
    """Per-component score breakdown for each threat type."""
    storm_id: str
    threat_scores: dict[str, float] = Field(default_factory=dict)
    component_breakdown: dict[str, dict[str, float]] = Field(default_factory=dict)
    top_reasons: dict[str, list[str]] = Field(default_factory=dict)
    limiting_factors: dict[str, list[str]] = Field(default_factory=dict)
    lifecycle_summary: dict = Field(default_factory=dict)
    provenance: str = (
        "Component scores are proxy-derived from radar-object fields and model environment data. "
        "They are not official operational scores."
    )


class EventFlagResponse(BaseModel):
    """A single operational event flag for a storm."""
    flag: str
    label: str
    confidence: float
    rationale: str
    severity: int
    provenance: str


class StormHotspotResponse(BaseModel):
    """A storm entry in the operational hotspots / priority ranking."""
    storm_id: str
    site: str
    priority_score: float
    priority_label: str
    severity_level: str
    primary_threat: str
    threat_scores: dict[str, float] = Field(default_factory=dict)
    storm_mode: str = "unknown"
    centroid_lat: float
    centroid_lon: float
    motion_heading_deg: float | None = None
    motion_speed_kmh: float | None = None
    confidence: float
    trend: str
    event_flags: list[EventFlagResponse] = Field(default_factory=list)
    top_flag: str | None = None
    impact_count: int = 0
    latest_scan_time: datetime | None = None


class LocationRiskEntry(BaseModel):
    """A saved location with its current storm threat intelligence."""
    location_id: str
    location_name: str
    lat: float
    lon: float
    risk_level: str          # HIGH / MODERATE / LOW / NONE
    risk_score: float        # 0–1
    threatening_storm_count: int
    nearest_eta_low: int | None = None
    nearest_eta_high: int | None = None
    primary_threat: str | None = None
    threat_scores: dict[str, float] = Field(default_factory=dict)
    top_storm_id: str | None = None
    top_storm_severity: str | None = None
    top_impact_summary: str | None = None
    trend: str | None = None
    confidence: float | None = None
    event_flag_labels: list[str] = Field(default_factory=list)


class StormCompareField(BaseModel):
    label: str
    storm_a: Any
    storm_b: Any
    delta: Any | None = None
    note: str | None = None


class StormCompareResponse(BaseModel):
    """Side-by-side storm comparison."""
    storm_a_id: str
    storm_b_id: str
    fields: list[StormCompareField] = Field(default_factory=list)
    provenance: str = "Comparison derived from current storm-object state. All metrics are proxy-derived."


class StormEventHistoryPoint(BaseModel):
    """Single scan-time snapshot of a storm's event flags and operational state."""
    scan_time: str
    event_flags: list[dict[str, Any]] = Field(default_factory=list)
    lifecycle_state: str | None = None
    priority_score: float | None = None
    priority_label: str | None = None
    severity_level: str | None = None
    primary_threat: str | None = None
    threat_scores: dict[str, float] = Field(default_factory=dict)
    storm_mode: str | None = None
    motion_heading_deg: float | None = None
    motion_speed_kmh: float | None = None
    confidence: float | None = None


class StormEventHistoryResponse(BaseModel):
    storm_id: str
    site: str
    point_count: int
    points: list[StormEventHistoryPoint] = Field(default_factory=list)
    provenance: str = "Event history persisted server-side per scan. All flags are proxy-derived heuristics."


class StormPrecomputedSummaryResponse(BaseModel):
    """Precomputed aggregated summary for a storm, built by the history aggregator."""
    storm_id: str
    site: str
    computed_at: str
    scan_count: int
    first_seen: str | None = None
    last_seen: str | None = None
    peak_severity: str | None = None
    peak_threat_scores: dict[str, float] = Field(default_factory=dict)
    peak_reflectivity: float | None = None
    max_area_km2: float | None = None
    max_speed_kmh: float | None = None
    max_priority_score: float | None = None
    dominant_mode: str | None = None
    flag_summary: list[dict[str, Any]] = Field(default_factory=list)
    threat_trend: list[dict[str, Any]] = Field(default_factory=list)
    motion_trend: list[dict[str, Any]] = Field(default_factory=list)
    impact_location_ids: list[str] = Field(default_factory=list)
    summary_narrative: str | None = None
    provenance: str = "Precomputed server-side by history aggregator. All metrics are proxy-derived."
